Complete rewrite of install docker.

Kangie opened the pull request:
- Add traefik template
- Restructure instructions to be easier to follow

~~Depends on LemmyNet/lemmy-ansible#124~~

Kangie commented:
Once this is merged it might be worthwhile to take the lemmy-ansible repo and make a lemmy-docker repo with a turnkey production-ish deployment (or two) that can be referenced in docs and checked out along with an appropriate `README.md`.

If you're willing to make an empty repo under LemmyNet I can submit a PR and get it ready-to-go!

Kangie commented:
FWIW,  the `dead links` CI/CD job doesn't do _anything_ at the moment. I'll push an update to both get it looking at the actual location of the source files _and_ doing online checks; though this _will_ break CI/CD - feel free not to merge the commit if that's a bother!

Nutomic commented:
Responded in https://github.com/LemmyNet/lemmy-ansible/pull/124#issuecomment-1632650401

Kangie commented:
@Nutomic; updated. It's failing CI because I fixed the dead_links CI job; but a quick look at the output _seems_ fine; just archive.org having load issues.

Nutomic commented on src/administration/install_docker.md:
```diff
@@ -1,78 +1,118 @@
+mkdir -p /data/lemmy/{volumes,config}
+mkdir -p /data/lemmy/volumes/{pictrs,postgres}
+chown 991:991 /data/lemmy/volumes/pictrs
+cd /data/lemmy
```
I would consider using the same dir as ansible for consistency which is `/srv/lemmy/{domain}`. And still mention that any path can be used.

Kangie replied:
I've implemented this, but would prefer that somewhere other than `/srv` be used - while `/srv` is part of the FHS not all Linuxes are FHS-compliant and Gentoo, for example, doesn't have a `/srv`. Any suggestions for alternatives?

As a bit of an aside this may be a good opportunity to simplify both guides - how many users are actually hosting multiple lemmy instances on a single system? I have enough trouble with the resource utilisation for a very small instance on a VPS...

My point is that any users  in a position to run multiple instances on a single server should have the prerequisite experience to know when they need to deviate from the provided instructions while the vast majority of users following this guide will not need anything like a multi-domain level of complexity.

Nutomic commented on src/administration/install_docker.md:
```diff
@@ -1,78 +1,118 @@
+
+## Updating
+
+To update lemmy (and other containers in the stack), change the version in `docker-compose.yml` then stop and start the stack.
```
You dont need to stop it, running `docker-compose up -d` is enough.

Nutomic commented on src/administration/install_docker.md:
```diff
@@ -1,78 +1,118 @@
+### TLS certificates
 
-`wget https://raw.githubusercontent.com/LemmyNet/lemmy-ansible/main/templates/nginx.conf`
+In 2023 the use of TLS is Mandatory. [Let's Encrypt](https://letsencrypt.org/) can be used to generate TLS certificates for the reverse proxy.
```
TLS is required for federation, the year doesnt matter.

dessalines commented on src/administration/install_docker.md:
```diff
@@ -1,78 +1,118 @@
-wget https://raw.githubusercontent.com/LemmyNet/lemmy-ansible/main/examples/config.hjson -O lemmy.hjson
-wget https://raw.githubusercontent.com/LemmyNet/lemmy-ansible/main/templates/nginx_internal.conf
+curl -O https://raw.githubusercontent.com/LemmyNet/lemmy-ansible/main/templates/docker-compose.yml
+pushd config
```
What's going on here? All these should be at the same level.

Kangie replied:
It really doesn't matter where the files get mounted in from; I thought it more explicit for documentation purposes to use a `./config` directory to bind in the configurations. We're already instructing users to pull and amend the `docker-compose.yml` template for Ansible for their non-Ansible deployments anyway...

dessalines commented on src/administration/install_docker.md:
```diff
@@ -1,78 +1,118 @@
+These templates contain `{{ }}` braces for variables, such as passwords and the domain name; edit them before starting up Lemmy for the first time.
 
-## Database tweaks
+#### Traefik
```
Remove this, we'll only support nginx. Traefik or the dozens of other potential servers should have instructions elsewhere.

Kangie replied:
> Remove this, we'll only support nginx. Traefik or the dozens of other potential servers should have instructions elsewhere.

I'd prefer not to do this. The entire rationale for revisiting these docs was to document Traefik as a load balancer / reverse proxy for Lemmy as it's been noted in several places as being ~~lacking~~ nonexistant; the docker documentation getting a touch-up is just a happy side-effect.

dessalines commented on src/administration/install_docker.md:
````diff
@@ -1,78 +1,118 @@
 docker-compose up -d
 ```
+
+`docker stack`:
````
docker stack? what is this? That isn't how to deploy lemmy.

Kangie replied:
`docker stack` is Docker's container orchestration tool for deployment to multi-node docker clusters. It provides better tooling to enable high availability and horizontal scaling. `docker compose` is for testing your multi-container deployments; stack is

Kangie replied:
docker stack is still an appropriate mechanism to deploy non-scaled services with the benefit that it each service can migrate within the swarm nodes in the event of e.g. a power failure. It should be documented for completeness sake, even if compose is the typical / preferred option.

Kangie commented:
Now that I've gone and put together the 'traefik lemmy docker` git repo what are your thoughts on an official `LemmyNet/Lemmy-Docker` repository that mirrors that structure but has `docker-compose.nginx.yml` and `docker-compose.traefik.yml` etc? It would be far simpler for users to get started and allow for a solid 1/2 of the shell instructions here to be removed in favour of `git clone ...`

Kangie commented:
@Nutomic any feedback at this stage?

I appreciate that you're both busy so I don't mind waiting a bit before this gets merged, just want to make sure that it doesn't fall off the RADAR.

dessalines reviewed (changes requested):
Your traefik and docker stack customizations should live in their own files. The rest of the PR just makes installation more verbose and geared towards custom webservers, when we will only support nginx.

Kangie commented on .woodpecker.yml:
```diff
@@ -38,4 +38,4 @@ pipeline:
   dead_links:
     image: becheran/mlc:latest
     commands:
```
What's your rationale for this, I don't care one way or the other and dropping the commit is trivial, but why keep it disabled?

Can I instead amend the commit to purge the CI that you're not using?

This is effectively 'dead code' prior to my change and I don't understand the value of treating CI configuration that you're _not_ using as 'live code' when it can just as easily (and if not in use frankly should) reside in the commit history.

dessalines replied:
We've had a lot of failed builds because of mlc (external links go down a lot).

dessalines commented:
Stale PR, but can re-open if necessary.
